doc/src/explanation/best-practices.md

```diff
@@ -0,0 +1,1329 @@
+  determine truth value of Relational`. The
+  {attr}`~sympy.core.expr.Expr.is_number` attribute can be used to check if an
+  expression can be evaluated to a specific number with `evalf()`. If you want
+  to only accept integers, you can check `isinstance(x, Integer)` (after
+  calling `sympify()` to convert Python ints). Beware that `is_integer` uses
+  the assumptions system and may be True even for symbolic objects, like
+  `Symbol('x', integer=True)`.
+
+- **Use the assumptions system.** If you do support symbolic inputs, you
+  should use the assumptions system to check for things like `x > 0`, e.g.,
+  using `x.is_positive`. When doing this, you should always [be aware of the
+  nuances](booleans-guide) of the {term}`three-valued fuzzy logic
+  <three-valued logic>` used in the assumptions system. That is, always be
+  aware that an assumption could be `None`, meaning its value is unknown and
+  could be either true or false. For example,
+
+  ```py
+  if x.is_positive:
+      ...
+  ```
+
+  will only run the block if `x.is_positive` is `True`, but you may want to do
+  something when `x.is_positive` is `None`.
+
+- **Return a Piecewise result.** If the result of a function depends on an
+  inequality or other boolean condition, you can use {class}`~.Piecewise` to
+  return a result that represents both possibilities symbolically. This is
+  generally preferred when possible, as it offers the most flexibility. This
+  is because the result is represented symbolically, meaning, for instance,
+  one can later substitute specific values for the symbols and it will
+  evaluate to the specific case, even if it is combined with other
+  expressions.
+
+  For example, instead of
+
+  ```py
+  if x > 0:
+      expr = 1
+  else:
+      expr = 0
+  ```
+
+  this can be represented symbolically as
+
+  ```py
+  >>> from sympy import Piecewise, pprint
+  >>> expr = Piecewise((1, x > 0), (0, True))
+  >>> pprint(expr, use_unicode=True)
+  ⎧1  for x > 0
+  ⎨
+  ⎩0  otherwise
+  >>> expr.subs(x, 1)
+  1
+  >>> expr.subs(x, -1)
+  0
+  ```
+
+- **Use {func}`~.ordered` to sort expressions into a canonical order.** If you
+  are trying to use `sorted` because you want a canonical ordering, but you
+  don't particularly care what that ordering is, you can use `ordered`.
+
+  ```py
+  >>> from sympy import ordered
+  >>> list(ordered([x, 0]))
+  [0, x]
+  ```
+
+  Alternatively, try to write the function in a way so that the correctness of
+  the result does not depend on the order that arguments are processed in.
+
+## Custom SymPy Objects
+
+SymPy is designed to be extended with custom classes, typically by subclassing
+{term}`Basic`, {term}`Expr`, or {term}`Function <Function (class)>`. All the
+symbolic classes in SymPy itself are written this way, and the points here
+apply equally to them as to user-defined classes.
+
+For an in-depth guide on how to write a `Function` subclass, see the [guide on
+writing custom functions](custom-functions).
```

Comment:
My opinion is SymPy rarely need to be extended though.

For example, I'm not sure about what kind of things are more capable to do with custom functions with `eval`, than `Function` with substitution)

Reply:
Both this guide and the one linked here apply equally well to user code and library code. Every class in SymPy is a "custom object".

Reply:
Basically, I wrote this section of this guide and the custom functions guide primarily for for SymPy developers, because we were sorely lacking in that sort of documentation and there are a lot of things involved. But everything there is public API and completely supported for users. It's true that the vast majority of users shouldn't be doing anything this advanced, but there are some who do, and there are legitimate reasons to do so. We should always aim to make the methods used to extend SymPy internally to also be usable by end-users. That way we keep the APIs for doing so clean and sufficiently extensible.